url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,555 @@
+        }
+        if (verifiedSignatures.sepSignatureRequired && !verifiedSignatures.sepSignaturePresent) {
+            result.add(new NoSecureEntryPointReason(q.name.ace));
+        }
+        if (!toBeVerified.isEmpty()) {
+            if (toBeVerified.size() != answers.size()) {
+                throw new DNSSECValidationFailedException(q, "Only some records are signed!");
+            } else {
+                result.add(new NoSignaturesReason(q));
+            }
+        }
+        return result;
+    }
+
+    private Set<UnverifiedReason> verifyNsec(DNSMessage dnsMessage) throws IOException {
+        Set<UnverifiedReason> result = new HashSet<>();
+        Question q = dnsMessage.questions.get(0);
+        boolean validNsec = false;
+        boolean nsecPresent = false;
+        DNSName zone = null;
+        List<Record<? extends Data>> nameserverRecords = dnsMessage.authoritySection;
+        for (Record<? extends Data> nameserverRecord : nameserverRecords) {
+            if (nameserverRecord.type == TYPE.SOA)
+                zone = nameserverRecord.name;
+        }
+        if (zone == null)
+            throw new DNSSECValidationFailedException(q, "NSECs must always match to a SOA");
+        for (Record<? extends Data> record : nameserverRecords) {
+            UnverifiedReason reason;
+
+            switch (record.type) {
+            case NSEC:
+                nsecPresent = true;
+                reason = verifier.verifyNsec(record, q);
+                break;
+            case NSEC3:
+                nsecPresent = true;
+                reason = verifier.verifyNsec3(zone, record, q);
+                break;
+            default:
+                continue;
+            }
+
+            if (reason != null) {
+                result.add(reason);
+            } else {
+                validNsec = true;
+            }
+        }
+        if (nsecPresent && !validNsec) {
+            throw new DNSSECValidationFailedException(q, "Invalid NSEC!");
+        }
+        List<Record<? extends Data>> toBeVerified = dnsMessage.copyAuthority();
+        VerifySignaturesResult verifiedSignatures = verifySignatures(q, nameserverRecords, toBeVerified);
+        if (validNsec && verifiedSignatures.reasons.isEmpty()) {
+            result.clear();
+        } else {
+            result.addAll(verifiedSignatures.reasons);
+        }
+        if (!toBeVerified.isEmpty() && toBeVerified.size() != nameserverRecords.size()) {
+            throw new DNSSECValidationFailedException(q, "Only some nameserver records are signed!");
+        }
+        return result;
+    }
+
+    private class VerifySignaturesResult {
+        boolean sepSignatureRequired = false;
+        boolean sepSignaturePresent = false;
+        Set<UnverifiedReason> reasons = new HashSet<>();
+    }
+
+    private VerifySignaturesResult verifySignatures(Question q, Collection<Record<? extends Data>> reference, List<Record<? extends Data>> toBeVerified) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 263-274, lines 276-283 into separate methods.